libc/socket/unix-socket.c

```diff
@@ -904,6 +928,587 @@ TEST(test_unix_socket, poll)
+
+		SEND_MSG(fd, pid, 0);
+
+		/* read msg from parent so that parent does the read
+		 * before conn gets closed */
+		READ_MSG(fd, pid, 0);
+		close(fd);
+
+		exit(0);
+	}
+
+	/* blocking connect, nonblocking accept */
+
+	pid = safe_fork();
+
+	if (pid) {
+		if ((named = unix_named_socket(type, socket_name)) < 0)
+			FAIL("unix_named_socket");
+
+		if (set_nonblock(named, 1) < 0)
+			FAIL("set_nonblock");
+
+		if (listen(named, 0) < 0)
+			FAIL("listen");
+
+		fds[0].fd = named;
+		fds[0].events = POLLIN;
+
+		TEST_ASSERT_EQUAL_INT(1, poll(fds, 1, 500));
+		TEST_ASSERT_EQUAL_INT(POLLIN, fds[0].revents);
+
+		conn = accept(fds[0].fd, NULL, NULL);
+		TEST_ASSERT(conn > 0);
+
+		TEST_ASSERT_EQUAL_INT(0, waitpid(pid, NULL, WNOHANG));
+
+		READ_MSG(conn, pid, 0);
+
+		SEND_MSG(conn, pid, 0);
+
+		if (waitpid(pid, &status, 0) < 0)
+			FAIL("waitpid");
+
+		TEST_ASSERT(WIFEXITED(status));
+		TEST_ASSERT_EQUAL_INT(0, WEXITSTATUS(status));
+
+		close(conn);
+		close(named);
+		unlink(socket_name);
+	}
+	else {
+		if ((fd = socket(AF_UNIX, type, 0)) < 0)
+			exit(1);
+
+		rv = connect_to_named_or_timeout(fd, socket_name, 3000);
+
+		SEND_MSG(fd, pid, 0);
+
+		READ_MSG(fd, pid, 0);
+		close(fd);
+
+		exit(0);
+	}
+
+	/* nonblocking connect, blocking accept */
+
+	pid = safe_fork();
+
+	if (pid) {
+		if ((fd = socket(AF_UNIX, type, 0)) < 0)
+			FAIL("socket");
+
+		if (set_nonblock(fd, 1) < 0)
+			FAIL("set_nonblock");
+
+		while (true) {
+			errno = 0;
+			rv = connect_to_named(fd, socket_name);
+			if (rv >= 0) {
+#ifdef __phoenix__
+				FAIL("should never happen - child proc should sleep for longer");
+#else
+				/* glibc behaves differently - see note in unix_accept_connect_errnos() */
+				break;
+#endif
+			}
+			else if (errno == EINPROGRESS) {
+				break;
+			}
+			else {
+				TEST_ASSERT(rv < 0);
+				TEST_ASSERT(errno == ECONNREFUSED || errno == ENOENT);
+				usleep(50);
+			}
+		}
+
+		fds[0].fd = fd;
+		fds[0].events = POLLOUT;
+
+		TEST_ASSERT_EQUAL_INT(1, poll(fds, 1, 700));
+		TEST_ASSERT_EQUAL_INT(POLLOUT, fds[0].revents);
+
+		int optval = 0;
+		int optlen = sizeof(optval);
+
+		TEST_ASSERT_EQUAL_INT(0, getsockopt(fds[0].fd, SOL_SOCKET, SO_ERROR, &optval, (socklen_t *)&optlen));
+
+		fds[0].events = POLLIN;
+		TEST_ASSERT_EQUAL_INT(1, poll(fds, 1, 250));
+		TEST_ASSERT_EQUAL_INT(POLLIN, fds[0].revents);
+
+		TEST_ASSERT_EQUAL_INT(0, waitpid(pid, NULL, WNOHANG));
+
+		READ_MSG(fds[0].fd, pid, 0);
+
+		fds[0].events = POLLOUT;
+		TEST_ASSERT_EQUAL_INT(1, poll(fds, 1, 250));
+		TEST_ASSERT_EQUAL_INT(POLLOUT, fds[0].revents);
+
+		SEND_MSG(fds[0].fd, pid, 0);
+
+		if (waitpid(pid, &status, 0) < 0)
+			FAIL("waitpid");
+
+		TEST_ASSERT(WIFEXITED(status));
+		TEST_ASSERT_EQUAL_INT(0, WEXITSTATUS(status));
+
+		close(fds[0].fd);
+	}
+	else {
+		if ((named = unix_named_socket(type, socket_name)) < 0)
+			exit(1);
+
+		if (listen(named, 0) < 0)
+			exit(1);
+
+		usleep(500); /* sleep so that connect would block */
```

Comment:
500 µs is too short, the test occasionally fails. Increase the sleep time to at least 1 ms.